### ISSUE TITLE
netty: Support address rewriting proxy mapper

### ISSUE THREAD

**Issue by ejona86** [grpc/grpc-java#10022]
[gRFC A1](https://github.com/grpc/proposal/blob/master/A1-http-connect-proxy-support.md) describes HTTP CONNECT support. grpc-java has supported "Use case 1" to one level or another since v1.0.3. There's been some ways to do "use case 2" (and 3), but these needed custom name resolvers and Netty ProtocolNegotiators for other reasons, and they could do what they needed without any direct support in grpc-java.

We're seeing some new cases for "use case 2." The important part of that case is the client is aware of backends so can load balance between them. Name resolution is generally not DNS, and may actually use a "use case 1" approach to get addresses.

We need to add an API, at least to Netty, to choose to use a proxy on a per-address basis. We can potentially use the ProxyDetector API for this, but we'd call it with resolved SocketAddresses instead of the normal unresolved addresses. I don't know if that's a good idea or not, also because ProxyDetector can block and we probably can't block in the usage here for Netty.

The implementation for this can be a Netty protocol negotiator that wraps the real negotiator if the builder was configured with one of these per-address mappers. It probably can reuse large portions of `ProtocolNegotiators.httpProxy()`.

**Comment by Jonpez2** [comment]
@ejona86 Thank you for this!  You mention prior art which addresses this in the netty layer rather than native to grpc; do you happen to have pointers to any examples?  It would be very useful to know of workarounds to help while this feature is being built.

**Comment by ejona86** [comment]
Looking at this more, I don't think we need a new protocol negotiatior. This can be done cross-transport as ClientTransportOptions already encodes whether a proxy should be used. Worst-case, the proxy mapping could be run in InternalSubchannel, but we could instead put the logic in ManagedChannelImpl or ManagedChannelImplBuilder by using a wrapping ClientTransportFactory.

**Comment by ejona86** [comment]
With the current API, you can hack something like the proxy mapper with:
```java
LoadBalancerRegistry registry = LoadBalancerRegistry.getDefaultRegistry();
// Some LB policy you are using
registry.register(new ProxyingLoadBalancerProvider(registry.getProvider("round_robin")));

// rewriteAddress() is the main interesting method below
final class ProxyingLoadBalancerProvider extends LoadBalancerProvider {
  private final LoadBalancerProvider delegate;

  public ProxyingLoadBalancerProvider(LoadBalancerProvider delegate) {
    if (delegate == null) {
      throw new NullPointerException("delegate");
    }
    this.delegate = delegate;
  }

  @Override public boolean isAvailable() { return delegate.isAvailable(); }
  @Override public int getPriority() { return delegate.getPriority() + 1; }
  @Override public String getPolicyName() { return delegate.getPolicyName(); }
  @Override public NameResolver.ConfigOrError parseLoadBalancingPolicyConfig(Map<String, ?> raw) {
    return delegate.parseLoadBalancingPolicyConfig(raw);
  }

  @Override
  public LoadBalancer newLoadBalancer(LoadBalancer.Helper helper) {
    return delegate.newLoadBalancer(new ProxyingHelper(helper));
  }

  private List<EquivalentAddressGroup> rewriteAddresses(List<EquivalentAddressGroup> eags) {
    List<EquivalentAddressGroup> newEags = new ArrayList<>(eags.size());
    for (EquivalentAddressGroup eag : eags) {
      List<SocketAddress> addrs = new ArrayList<>(eag.getAddresses().size());
      for (SocketAddress addr : eag.getAddresses()) {
        addrs.add(rewriteAddress(addr));
      }
      newEags.add(new EquivalentAddressGroup(addrs, eag.getAttributes()));
    }
    return newEags;
  }

  /** THIS IS ESSENTIALLY THE PROXY MAPPER */
  private SocketAddress rewriteAddress(SocketAddress addr) {
    if (!(addr instanceof InetSocketAddress)) {
      return addr;
    }
    InetSocketAddress inet = (InetSocketAddress) addr;
    InetAddress newAddr;
    try {
      // Strip any hostname
      newAddr = InetAddress.getByAddress(inet.getAddress().getAddress());
    } catch (UnknownHostException ex) {
      throw new AssertionError(ex); // Not possible
    }
    addr = HttpConnectProxiedSocketAddress.newBuilder()
        .setTargetAddress(new InetSocketAddress(newAddr, inet.getPort()))
        .setProxyAddress(new InetSocketAddress("127.0.0.1", 1234))
        .build();
    return addr;
  }

  private final class ProxyingHelper extends io.grpc.util.ForwardingLoadBalancerHelper {
    private final LoadBalancer.Helper delegate;

    public ProxyingHelper(LoadBalancer.Helper delegate) {
      this.delegate = delegate;
    }

    @Override protected LoadBalancer.Helper delegate() { return delegate; }

    @Override
    public LoadBalancer.Subchannel createSubchannel(LoadBalancer.CreateSubchannelArgs args) {
      return new ProxyingSubchannel(
          delegate.createSubchannel(
              args.toBuilder().setAddresses(rewriteAddresses(args.getAddresses())).build()),
          args.getAddresses());
    }
  }

  private final class ProxyingSubchannel extends io.grpc.util.ForwardingSubchannel {
    private final LoadBalancer.Subchannel delegate;
    private List<EquivalentAddressGroup> addresses;

    public ProxyingSubchannel(
        LoadBalancer.Subchannel delegate, List<EquivalentAddressGroup> addresses) {
      this.delegate = delegate;
      this.addresses = addresses;
    }

    @Override protected LoadBalancer.Subchannel delegate() { return delegate; }
    @Override public List<EquivalentAddressGroup> getAllAddresses() { return addresses; }

    @Override
    public void updateAddresses(List<EquivalentAddressGroup> addrs) {
      super.updateAddresses(rewriteAddresses(addrs));
      addresses = new ArrayList<>(addrs);
    }
  }
}
```

**Comment by becomeStar** [comment]
Hi @YifeiZhuang, @ejona86,

I’ve been looking into this issue and would like to help move forward **per-address proxy mapping** (Use Case 2 of gRFC A1). Since there hasn’t been much recent activity, I put together a working prototype to validate the approach and surface concrete design trade-offs.

I’m proposing a **two-phased approach**, where Phase 1 is explicitly **verification-only**, and Phase 2 focuses on the final integration point.

---

### Phase 1 – Internal Verification (LB-level Prototype)

In Phase 1, I implemented a `ProxyingLoadBalancerProvider` as a **decorator** around an existing LB policy (configured to wrap `round_robin` by default).

The goal of this phase is **not** to define the final integration point, but to validate that:

* Per-address proxy mapping behaves correctly at runtime
* Address rewriting integrates cleanly with existing LB policies
* Potentially blocking proxy detection can be moved out of subchannel creation

Key characteristics:

* Proxy resolution is **pre-computed during `handleResolvedAddresses()`**, avoiding blocking work in the `SynchronizationContext`
* A **bounded, access-order LRU cache** is used to prevent unbounded memory growth
* Exposed as a temporary, standalone policy (`proxying`) **purely for verification and testing**
* Includes unit tests covering address rewriting, proxy detection behavior, and cache semantics

This phase is intentionally scoped as an **internal engine and proof-of-concept**, not a committed API or registry design.

---

### Phase 2 – Public API & Final Integration

Once the mechanism is validated, the intent is to move the logic to the **appropriate channel-level integration point**, rather than keeping it as an LB decorator.

Based on prior discussion and feedback, the current direction for Phase 2 would be:

* Allow users to provide a custom `ProxyMapper` via `ManagedChannelImplBuilder`
* Propagate the mapper internally (e.g., via `Attributes.Key` in `ResolvedAddresses`, or equivalently through a wrapped `ClientTransportFactory`)
* Apply proxy mapping only when a custom mapper is present
* Preserve **zero overhead for default configurations**
* Avoid introducing new entries into `LoadBalancerRegistry` or altering policy selection behavior

In other words, Phase 1 validates *that* per-address proxy mapping works correctly, while Phase 2 determines *where* it belongs architecturally.

---

If this phased approach aligns with the project’s expectations, I’d be happy to prepare a Phase 1 PR for review, clearly marked as verification-only and not intended as the final integration.

Feedback on the direction — especially around the preferred integration point for Phase 2 — would be greatly appreciated.

I'm also open to going directly to a Phase 2 design if the direction already feels clear and a separate Phase 1 validation step wouldn't add much value.

**Comment by ejona86** [comment]
Support for this has been added to xDS, so the original need for it as an API has gone away to my knowledge. So it is probably better to hold of on implementing this until it is needed again (to make sure the implementation actually fits the requirements).

I don't think we would need phase 1. We aren't going to have a cache and whether the behavior works is mostly dictated by the ProxyDetector implementation. There is no need to propagate the mapper through `Attributes.Key`; it can be passed directly to all the places it might be used.

**Comment by becomeStar** [comment]
@ejona86 

Thanks for the clarification — that makes sense.
If xDS already covers the primary use cases today, I agree it’s better to hold off until a concrete non-xDS requirement emerges.

It’s also helpful to know that in a future design, this would likely live at the channel / transport layer (without caching, Phase 1, or Attributes-based propagation). I’ll keep that in mind if the need comes up again.

Appreciate the guidance.

**Comment by ejona86** [comment]
This could be done if someone needs it, but closing since it seems nobody needs it for the while.
